File: typescript/src/integrations/openai.ts
/**
 * OpenAI Integration — auto-orchestrate OpenAI tool calls via Synapse.
 *
 * @example
 * ```ts
 * import OpenAI from "openai";
 * import { SynapseOpenAI } from "synapse-orchestrator/integrations/openai";
 *
 * const client = new SynapseOpenAI({
 *   openaiClient: new OpenAI(),
 *   tools: {
 *     getWeather:  getWeatherFn,
 *     getForecast: getForecastFn,
 *     sendAlert:   sendAlertFn,
 *   },
 * });
 *
 * const [response, reports] = await client.chat({
 *   model: "gpt-4o",
 *   messages: [{ role: "user", content: "What's the weather in NYC?" }],
 *   tools: openAIToolSchemas,
 * });
 * ```
 */

import { Orchestrator } from "../orchestrator.js";
import type { AsyncToolFn, ExecutionReport } from "../executor.js";
import type { ToolCall } from "../dependencyAnalyzer.js";
import type { OrchestratorOptions } from "../orchestrator.js";

// ── Minimal OpenAI type shims (avoids a hard dep on the openai package) ───────

interface OpenAIToolCall {
  id: string;
  function: { name: string; arguments: string };
}

interface OpenAIMessage {
  tool_calls?: OpenAIToolCall[];
  [key: string]: unknown;
}

interface OpenAIChoice {
  message: OpenAIMessage;
}

interface OpenAIResponse {
  choices: OpenAIChoice[];
}

interface OpenAIClient {
  chat: {
    completions: {
      create(params: Record<string, unknown>): Promise<OpenAIResponse>;
    };
  };
}

// ── Helpers ────────────────────────────────────────────────────────────────────

function openAICallsToSynapse(toolCalls: OpenAIToolCall[]): ToolCall[] {
  return toolCalls.map((tc) => ({
    id: tc.id,
    name: tc.function.name,
    inputs: JSON.parse(tc.function.arguments) as Record<string, unknown>,
  }));
}

function buildToolMessages(report: ExecutionReport): Array<Record<string, unknown>> {
  const messages: Array<Record<string, unknown>> = [];
  for (const [callId, result] of report.results) {
    messages.push({
      role: "tool",
      tool_call_id: callId,
      content:
        result.status === "success"
          ? JSON.stringify(result.output)
          : JSON.stringify({ error: result.error ?? "Unknown tool error", status: result.status }),
    });
  }
  return messages;
}

// ── Main class ─────────────────────────────────────────────────────────────────

export interface SynapseOpenAIOptions extends Omit<OrchestratorOptions, "tools"> {
  openaiClient: OpenAIClient;
  tools: Record<string, AsyncToolFn>;
  maxRounds?: number;
}

export class SynapseOpenAI {
  private client: OpenAIClient;
  private orch: Orchestrator;
  private maxRounds: number;

  constructor(options: SynapseOpenAIOptions) {
    this.client = options.openaiClient;
    this.orch = new Orchestrator({ ...options });
    this.maxRounds = options.maxRounds ?? 10;
  }

  async chat(params: {
    model: string;
    messages: Array<Record<string, unknown>>;
    tools: Array<Record<string, unknown>>;
    [key: string]: unknown;
  }): Promise<[OpenAIResponse, ExecutionReport[]]> {
    const { model, messages, tools, ...rest } = params;
    const history = [...messages];
    const allReports: ExecutionReport[] = [];

    for (let round = 0; round < this.maxRounds; round++) {
      const response = await this.client.chat.completions.create({
        model,
        messages: history,
        tools,
        ...rest,
      });

      const message = response.choices[0].message;

      if (!message.tool_calls || message.tool_calls.length === 0) {
        return [response, allReports];
      }

      history.push(message as Record<string, unknown>);

      const synapseCalls = openAICallsToSynapse(message.tool_calls);
      const report = await this.orch.run(synapseCalls);
      allReports.push(report);

      const toolMessages = buildToolMessages(report);
      history.push(...toolMessages);
    }

    throw new Error(`Exceeded maxRounds=${this.maxRounds} without a final response.`);
  }
}

File: python/synapse/integrations/openai.py
"""
OpenAI Integration — auto-orchestrate OpenAI function/tool calls.

Drop-in wrapper around the OpenAI client that intercepts tool call
responses, builds a Synapse execution plan, and executes calls in
the optimal order before returning results back to the conversation.

Usage
-----
::

    from openai import AsyncOpenAI
    from synapse.integrations.openai import SynapseOpenAI

    client = SynapseOpenAI(
        openai_client=AsyncOpenAI(),
        tools={
            "get_weather":  get_weather,
            "get_forecast": get_forecast,
            "send_alert":   send_alert,
        },
    )

    response, report = await client.chat(
        model="gpt-4o",
        messages=[{"role": "user", "content": "What's the weather and 5-day forecast for NYC?"}],
        tools=openai_tool_schemas,
    )

    print(report)   # shows parallelism, speedup, per-call timing
"""

from __future__ import annotations

import json
from typing import Any

from ..dependency_analyzer import ToolCall
from ..executor import AsyncToolFn, ExecutionReport
from ..orchestrator import Orchestrator


def _openai_calls_to_synapse(tool_calls: list[Any]) -> list[ToolCall]:
    """
    Convert a list of OpenAI ChatCompletionMessageToolCall objects into
    Synapse ToolCall objects.

    Synapse performs its own dependency detection via $results.* references
    in inputs — if the model embeds such references in its arguments,
    they will be honoured automatically.
    """
    return [
        ToolCall(
            id=tc.id,
            name=tc.function.name,
            inputs=json.loads(tc.function.arguments),
        )
        for tc in tool_calls
    ]


def _build_tool_messages(report: ExecutionReport) -> list[dict[str, Any]]:
    """
    Build the list of role=tool messages to return to the model after
    all tool calls have been executed.
    """
    messages = []
    for call_id, result in report.results.items():
        content = (
            json.dumps(result.output)
            if result.output is not None
            else json.dumps({"error": result.error})
        )
        messages.append(
            {
                "role": "tool",
                "tool_call_id": call_id,
                "content": content,
            }
        )
    return messages


class SynapseOpenAI:
    """
    Wraps an AsyncOpenAI client and intercepts tool call responses so
    Synapse can execute them in parallel.

    Parameters
    ----------
    openai_client:
        An ``openai.AsyncOpenAI`` instance (caller owns lifecycle).
    tools:
        Mapping from tool name → async callable.
    max_rounds:
        Maximum agentic rounds (model call → tool execution → …).
    orchestrator_kwargs:
        Keyword arguments forwarded to ``Orchestrator.__init__``.
    """

    def __init__(
        self,
        openai_client: Any,
        tools: dict[str, AsyncToolFn],
        *,
        max_rounds: int = 10,
        **orchestrator_kwargs: Any,
    ) -> None:
        self._client = openai_client
        self._orch = Orchestrator(tools, **orchestrator_kwargs)
        self.max_rounds = max_rounds

    async def chat(
        self,
        *,
        model: str,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]],
        **kwargs: Any,
    ) -> tuple[Any, list[ExecutionReport]]:
        """
        Run a full agentic loop with automatic parallel tool execution.

        Returns
        -------
        (final_response, reports):
            ``final_response`` is the last ChatCompletion where the model
            produced text rather than tool calls.
            ``reports`` is one ExecutionReport per round of tool calls.
        """
        history = list(messages)
        all_reports: list[ExecutionReport] = []

        for _ in range(self.max_rounds):
            response = await self._client.chat.completions.create(
                model=model,
                messages=history,
                tools=tools,
                **kwargs,
            )
            message = response.choices[0].message

            if not message.tool_calls:
                # Model is done — return the final response
                return response, all_reports

            history.append(message)

            synapse_calls = _openai_calls_to_synapse(message.tool_calls)
            report = await self._orch.run(synapse_calls)
            all_reports.append(report)

            tool_messages = _build_tool_messages(report)
            history.extend(tool_messages)

        raise RuntimeError(
            f"Exceeded max_rounds={self.max_rounds} without a final response."
        )
